nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,567 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  isPath = x: builtins.substring 0 1 x == "/";
+ 
+  mkExcludeFile = cfg:
+    # Write each exclude pattern to a new line
+    pkgs.writeText "excludefile" (concatStringsSep "\n" cfg.exclude);
+
+  mkKeepArgs = cfg:
+    # If cfg.prune.keep e.g. has a yearly attribute,
+    # its content is passed on as --keep-yearly
+    concatStringsSep " "
+      (mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
+
+  mkBackupScript = cfg: ''
+    on_exit()
+    {
+      exit_status=$?
+      # Reset the EXIT handler, or else we're called again on 'exit' below
+      trap - EXIT
+      ${cfg.postHook}
+      exit $exit_status
+    }
+    trap 'on_exit' INT TERM QUIT EXIT
+
+    archiveName="$archiveName-$(date ${cfg.dateFormat})"
+    ${cfg.preHook}
+  '' + optionalString cfg.doInit ''
+    # Run borg init if the repo doesn't exist yet
+    if ! borg list > /dev/null; then
+      borg init \
+        --encryption ${cfg.encryption.mode} \
+        $extraInitArgs
+      ${cfg.postInit}
+    fi
+  '' + ''
+    archiveSuffix="${optionalString cfg.appendFailedSuffix ".failed"}"
+    borg create \
+      --compression ${cfg.compression} \
+      --exclude-from ${mkExcludeFile cfg} \
+      $extraCreateArgs \
+      "::$archiveName$archiveSuffix" \
+      ${escapeShellArgs cfg.paths}
+  '' + optionalString cfg.appendFailedSuffix ''
+    # --exit-status makes jq fail if .archive.name is not found
```

Comment:
The above comment can be removed (jq isn't used anymore).

Reply:
done